OCLM-30 Implement concept and mapping version check at importing

PawelGutkowski opened the pull request.

AdamGrzybkowski commented on api/src/test/java/org/openmrs/module/openconceptlab/updater/ImporterTest.java:
```diff
@@ -20,6 +20,7 @@
 import static org.hamcrest.Matchers.not;
 import static org.hamcrest.Matchers.notNullValue;
 import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNotNull;
```
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Avoid unused imports such as 'org.junit.Assert.assertNotNull'](https://www.codacy.com/app/agrzybkowski/openmrs-module-openconceptlab/file/2377174879/issues/source?bid=3218637&fileBranchId=3224807#l23)

rkorytkowski replied:
@AdamGrzybkowski, did you experiment with codacy here? I wonder how this comment ended up here?

AdamGrzybkowski replied:
Yes, I did before Paweł but I didn't have time to add coverage and then  Paweł did that. 
I have deleted this project from my codacy account today so it shoudn't add any  comments

AdamGrzybkowski commented on api/src/test/java/org/openmrs/module/openconceptlab/updater/ImporterTest.java:
```diff
@@ -464,6 +468,47 @@ public void importConcept_shouldChangeDuplicateSynonymToIndexTerm() throws Excep
+	 * @verifies save concept if versionUrl changed from last update
+	 */
+	@Test
+	public void importConcept_shouldUpdateConceptIfVersionUrlChanged() throws Exception {
```
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Method names should not contain underscores](https://www.codacy.com/app/agrzybkowski/openmrs-module-openconceptlab/file/2377174879/issues/source?bid=3218637&fileBranchId=3224807#l477)

AdamGrzybkowski commented on api/src/test/java/org/openmrs/module/openconceptlab/updater/ImporterTest.java:
```diff
@@ -671,6 +716,77 @@ public void importMapping_addConceptMappingAndUnretireTerm() throws Exception {
+		assertTrue(importer.isMappingUpdatedSince(item, oclMapping));
+	}
+	@Test
+	public void isMappingUpdatedSince_shouldReturnFalseIfBothUpdatedOnAreNull() throws Exception{
```
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Method names should not contain underscores](https://www.codacy.com/app/agrzybkowski/openmrs-module-openconceptlab/file/2377174879/issues/source?bid=3218637&fileBranchId=3224807#l779)

AdamGrzybkowski commented on api/src/test/java/org/openmrs/module/openconceptlab/updater/ImporterTest.java:
```diff
@@ -671,6 +716,77 @@ public void importMapping_addConceptMappingAndUnretireTerm() throws Exception {
+	}
+	
+	@Test
+	public void isMappingUpdatedSince_shouldReturnIfMappingUpdateOnIsAfter() throws Exception{
```
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Method names should not contain underscores](https://www.codacy.com/app/agrzybkowski/openmrs-module-openconceptlab/file/2377174879/issues/source?bid=3218637&fileBranchId=3224807#l747)

AdamGrzybkowski commented on api/src/test/java/org/openmrs/module/openconceptlab/updater/ImporterTest.java:
```diff
@@ -464,6 +468,47 @@ public void importConcept_shouldChangeDuplicateSynonymToIndexTerm() throws Excep
+	 * @verifies skip saving concept if versionUrl didn't change from last update
+	 */
+	@Test
+	public void importConcept_shouldSkipUpdatingConceptIfVersionUrlDidntChange() throws Exception {
```
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Method names should not contain underscores](https://www.codacy.com/app/agrzybkowski/openmrs-module-openconceptlab/file/2377174879/issues/source?bid=3218637&fileBranchId=3224807#l498)

AdamGrzybkowski commented on api/src/test/java/org/openmrs/module/openconceptlab/updater/ImporterTest.java:
```diff
@@ -671,6 +716,77 @@ public void importMapping_addConceptMappingAndUnretireTerm() throws Exception {
+	}
+	
+	@Test
+	public void isMappingUpdatedSince_shouldReturnTrueIfItemUpdateOnIsNull() throws Exception{
```
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Method names should not contain underscores](https://www.codacy.com/app/agrzybkowski/openmrs-module-openconceptlab/file/2377174879/issues/source?bid=3218637&fileBranchId=3224807#l765)

AdamGrzybkowski commented on api/src/test/java/org/openmrs/module/openconceptlab/updater/ImporterTest.java:
```diff
@@ -671,6 +716,77 @@ public void importMapping_addConceptMappingAndUnretireTerm() throws Exception {
 	}
+	
+	@Test
+	public void importMapping_shouldUpdateMappingOnylIfItHasBeenUpdatedSinceLastImport() throws Exception {
```
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Method names should not contain underscores](https://www.codacy.com/app/agrzybkowski/openmrs-module-openconceptlab/file/2377174879/issues/source?bid=3218637&fileBranchId=3224807#l721)

AdamGrzybkowski commented on api/src/main/java/org/openmrs/module/openconceptlab/updater/Importer.java:
```diff
@@ -9,10 +9,15 @@
 import java.util.List;
+import java.util.TimeZone;
 
+import org.antlr.runtime.tree.RewriteRuleNodeStream;
```
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Avoid unused imports such as 'org.antlr.runtime.tree.RewriteRuleNodeStream'](https://www.codacy.com/app/agrzybkowski/openmrs-module-openconceptlab/file/2377174873/issues/source?bid=3218637&fileBranchId=3224807#l20)

AdamGrzybkowski commented on api/src/main/java/org/openmrs/module/openconceptlab/updater/Importer.java:
```diff
@@ -250,8 +261,14 @@ private void changeDuplicateNamesToIndexTerms(Concept concept, List<String> reso
 	public Item importMapping(CacheService cacheService, Update update, OclMapping oclMapping) {
+		Item oldMappingItem = updateService.getLastSuccessfulItemByUrl(oclMapping.getUrl());
+		if(oldMappingItem!= null && !isMappingUpdatedSince(oldMappingItem, oclMapping)){
+			return new Item(update, oclMapping, ItemState.ALREADY_UP_TO_DATE);
```
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [An empty statement (semicolon) not part of a loop](https://www.codacy.com/app/agrzybkowski/openmrs-module-openconceptlab/file/2377174873/issues/source?bid=3218637&fileBranchId=3224807#l270)

AdamGrzybkowski commented on api/src/main/java/org/openmrs/module/openconceptlab/updater/Importer.java:
```diff
@@ -250,8 +261,14 @@ private void changeDuplicateNamesToIndexTerms(Concept concept, List<String> reso
 	 * @should add concept mapping and unretire term
 	 * @should remove concept mapping and retire term
+	 * @should update mapping only if it has been updated since last import
 	 */
```
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Avoid really long methods.](https://www.codacy.com/app/agrzybkowski/openmrs-module-openconceptlab/file/2377174873/issues/source?bid=3218637&fileBranchId=3224807#l266)

rkorytkowski commented on api/src/main/resources/liquibase.xml:
```diff
@@ -131,5 +131,15 @@
+
+    <changeSet author="pawelgutkowski" id="openconceptlab-v1-11">
+    	<preConditions onFail="MARK_RAN">
+    		<not><columnExists indexName="openconceptlab_item_updated_on"/></not>
```
Probably columnName and not indexName. Please try installing the module in openmrs to see if the liquibase changeset is applied without errors (module won't start unless it can be applied)

rkorytkowski commented on api/src/main/resources/liquibase.xml:
```diff
@@ -131,5 +131,15 @@
+    	</preConditions>
+    	<comment>Create openconceptlab_item_update_on</comment>
+    	<addColumn tableName="openconceptlab_item">
+    		<column name="openconceptlab_updated_on" type="varchar(255)"/>
```
Probably you want to store it as date.

rkorytkowski commented on api/src/main/java/org/openmrs/module/openconceptlab/Item.java:
```diff
@@ -71,6 +71,10 @@
 	
+	@Basic
+	@Column(name = "updated_on")
+	private String updatedOn;
```
You want to store it as date, not string.

rkorytkowski commented on api/src/main/java/org/openmrs/module/openconceptlab/client/OclMapping.java:
```diff
@@ -45,6 +45,9 @@
 	private String toConceptCode;
 	
+	@JsonProperty("updated_on")
+	private String updatedOn;
```
You could try declaring is as Date and see if it is correctly converted from json.

PawelGutkowski commented:
I've changed type to Date, ObjectMapper is configured to parse date format. I've fixed POM so there is no warning messages that coberture version is not declared.

rkorytkowski commented:
Solid work Pawel! Thanks!
